File: antia_numerical_methods/c/fun/splint.c
/*	To compute integral of a tabulated function using cubic splines

	A : (input) Lower limit of integration
	B : (input) Upper limit of integration (B > A)
	SINT : (output) Computed value of integral using cubic splines
	TINT : (output) Computed value of integral using trapezoidal rule
	N : (input) Number of tabular points
	X : (input) Array of length N containing abscissas (in ascending order)
	F : (input) Array of length N containing the function values
	C : (input) Array of length 3*N containing the coefficients of splines
		
	Error status is returned by the value of the function SPLINT.
		0 value implies successful execution
		31 implies lower limit A is outside the table
		32 implies upper limit B is outside the table
		301 implies A>B or X[0]>X[N-1] and no calculations are done
		Other values may be set by SPLEVL

	Required functions : SPLEVL
*/

#include <math.h>


double splevl(double xb, int n, double x[], double f[], double c[][3],
		double *dfb, double *ddfb, int *ier);

int splint(double a, double b, double *sint, double *tint, int n, double x[],
	double f[], double c[][3])

{
	int i,ier,ier1;
	double a1,f1,f2,dfb,ddfb,b1,d1,d2;

	*sint=0.0; *tint=0.0;
	if(a == b) return 0;
	if(a>b || x[0]>=x[n-1]) return 301;

	a1=a;
/*	Evaluate the function at x=A for trapezoidal rule */
	f1=splevl(a1,n,x,f,c,&dfb,&ddfb,&ier);
	if(ier>100) return ier;
	if(a<x[0] || a>x[n-1]) ier=31;
	if(b<x[0] || b>x[n-1]) ier=32;

/*	Integrating over the n-1 subintervals */
	for(i=0; i<n-1; ++i) {
		if(a1 < x[i+1] || i == n-2) {
			b1=x[i+1]; if(b<b1) b1=b;
			if(i == n-2) b1=b;
			d1=a1-x[i];
			d2=b1-x[i];

/*	Add integral of cubic spline over [A1,B1] */
			*sint = *sint+f[i]*(d2-d1)+c[i][0]*(d2*d2-d1*d1)/2.+
				c[i][1]*(d2*d2*d2-d1*d1*d1)/3.+
				c[i][2]*(d2*d2*d2*d2-d1*d1*d1*d1)/4.;
			f2=f[i+1];
			if(b1 != x[i+1]) f2=splevl(b1,n,x,f,c,&dfb,&ddfb,&ier1);

/*	Trapezoidal rule approximation to integral */
			*tint = *tint+0.5*(b1-a1)*(f1+f2);
			if(b<=x[i+1]) return ier;
			a1=b1;
			f1=f2;
		}
	}
	return ier;
}

File: antia_numerical_methods/f77/sub4/splint.f
!	To compute integral of a tabulated function using cubic splines
!
!	A : (input) Lower limit of integration
!	B : (input) Upper limit of integration (B > A)
!	SINT : (output) Computed value of integral using cubic splines
!	TINT : (output) Computed value of integral using trapezoidal rule
!	N : (input) Number of tabular points
!	X : (input) Array of length N containing abscissas (in ascending order)
!	F : (input) Array of length N containing the function values
!	C : (input) Array of length 3*N containing the coefficients of splines
!	IER : (output) Error parameter, IER=0 implies successful execution
!		IER=31 implies lower limit A is outside the table
!		IER=32 implies upper limit B is outside the table
!		IER=301 implies A>B or X(1)>X(N) and no calculations are done
!		Other values may be set by SPLEVL
!
!	Required routines : SPLEVL
!
      SUBROUTINE SPLINT(A,B,SINT,TINT,N,X,F,C,IER)
!      IMPLICIT REAL*8(A-H,O-Z)
      DIMENSION X(N),F(N),C(3,N)

      IER=0
      SINT=0.0
      TINT=0.0
      IF(A.EQ.B) RETURN
      IER=301
      IF(A.GT.B.OR.X(1).GE.X(N)) RETURN

      A1=A
!	Evaluate the function at x=A for trapezoidal rule
      F1=SPLEVL(A1,N,X,F,C,DFB,DDFB,IER)
      IF(IER.GT.100) RETURN
      IER=0
      IF(A.LT.X(1).OR.A.GT.X(N)) IER=31
      IF(B.LT.X(1).OR.B.GT.X(N)) IER=32

!	Integrating over the N-1 subintervals
      DO 2000 I=1,N-1
        IF(A1.LT.X(I+1).OR.I.EQ.N-1) THEN
          B1=MIN(B,X(I+1))
          IF(I.EQ.N-1) B1=B
          D1=A1-X(I)
          D2=B1-X(I)
!	Add integral of cubic spline over [A1,B1]
          SINT=SINT+F(I)*(D2-D1)+C(1,I)*(D2*D2-D1*D1)/2.+
     1         C(2,I)*(D2**3-D1**3)/3.+C(3,I)*(D2**4-D1**4)/4.
          F2=F(I+1)
          IF(B1.NE.X(I+1)) F2=SPLEVL(B1,N,X,F,C,DFB,DDFB,IER1)
!	Trapezoidal rule approximation to integral
          TINT=TINT+0.5*(B1-A1)*(F1+F2)
          IF(B.LE.X(I+1)) RETURN
          A1=B1
          F1=F2
        ENDIF
2000  CONTINUE
      END
